transaction/beef.go

```diff
@@ -171,3 +288,587 @@ func (t *Transaction) collectAncestors(txns map[string]*Transaction) ([]string,
+		return tx
+	}
+	return nil
+}
+
+func (b *Beef) MakeTxidOnly(txid string) *BeefTx {
+	tx, ok := b.Transactions[txid]
+	if !ok {
+		return nil
+	}
+	if tx.DataFormat == TxIDOnly {
+		return tx
+	}
+	delete(b.Transactions, txid)
+	tx = &BeefTx{
+		DataFormat: TxIDOnly,
+		KnownTxID:  tx.KnownTxID,
+	}
+	b.Transactions[txid] = tx
+	return tx
+}
+
+func (b *Beef) MergeRawTx(rawTx []byte, bumpIndex *int) (*BeefTx, error) {
+	tx := &Transaction{}
+	reader := bytes.NewReader(rawTx)
+	_, err := tx.ReadFrom(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	txid := tx.TxID().String()
+	b.RemoveExistingTxid(txid)
+
+	beefTx := &BeefTx{
+		DataFormat:  RawTx,
+		Transaction: tx,
+	}
+
+	if bumpIndex != nil {
+		if *bumpIndex < 0 || *bumpIndex >= len(b.BUMPs) {
+			return nil, fmt.Errorf("invalid bump index")
+		}
+		beefTx.Transaction.MerklePath = b.BUMPs[*bumpIndex]
+		beefTx.DataFormat = RawTxAndBumpIndex
+	}
+
+	b.Transactions[txid] = beefTx
+	b.tryToValidateBumpIndex(beefTx)
+
+	return beefTx, nil
+}
+
+// RemoveExistingTxid removes an existing transaction from the BEEF, given its TXID
+func (b *Beef) RemoveExistingTxid(txid string) {
+	delete(b.Transactions, txid)
+}
+
+func (b *Beef) tryToValidateBumpIndex(tx *BeefTx) {
+	if tx.Transaction.MerklePath == nil {
+		return
+	}
+	for _, node := range tx.Transaction.MerklePath.Path[0] {
+		if node.Hash.String() == tx.Transaction.TxID().String() {
+			return
+		}
+	}
+	tx.Transaction.MerklePath = nil
+}
+
+func (b *Beef) MergeTransaction(tx *Transaction) (*BeefTx, error) {
+	txid := tx.TxID().String()
+	b.RemoveExistingTxid(txid)
+
+	var bumpIndex *int
+	if tx.MerklePath != nil {
+		index := b.MergeBump(tx.MerklePath)
+		bumpIndex = &index
+	}
+
+	newTx := &BeefTx{
+		DataFormat:  RawTx,
+		Transaction: tx,
+	}
+	if bumpIndex != nil {
+		newTx.DataFormat = RawTxAndBumpIndex
+	}
+
+	b.Transactions[txid] = newTx
+	b.tryToValidateBumpIndex(newTx)
+
+	if bumpIndex == nil {
+		for _, input := range tx.Inputs {
+			if input.SourceTransaction != nil {
+				if _, err := b.MergeTransaction(input.SourceTransaction); err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+
+	return newTx, nil
+}
+
+func (b *Beef) MergeTxidOnly(txid string) *BeefTx {
+	tx := b.findTxid(txid)
+	if tx == nil {
+		tx = &BeefTx{
+			DataFormat: TxIDOnly,
+			KnownTxID:  &chainhash.Hash{},
+		}
+		copy(tx.KnownTxID[:], txid)
+		b.Transactions[txid] = tx
+		b.tryToValidateBumpIndex(tx)
+	}
+	return tx
+}
+
+func (b *Beef) MergeBeefTx(btx *BeefTx) (*BeefTx, error) {
+	beefTx := b.findTxid(btx.Transaction.TxID().String())
+	if btx.DataFormat == TxIDOnly && beefTx == nil {
+		beefTx = b.MergeTxidOnly(btx.KnownTxID.String())
+	} else if btx.Transaction != nil && (beefTx == nil || beefTx.DataFormat == TxIDOnly) {
+		var err error
+		beefTx, err = b.MergeTransaction(btx.Transaction)
+		if err != nil {
+			return nil, err
+		}
+		// } else if btx.Transaction != nil && (beefTx == nil || beefTx.DataFormat == TxIDOnly) {
```

Comment:
@sirdeggen this appears to be doing the same check twice, such that the 2nd one would never be hit. I commented it out to fix the linter, but I wanted you to review before I just deleted it in case you meant something else here